### ISSUE TITLE
HomeConnect API unexpected values (WAVH8E41CH/07)

### ISSUE THREAD

**Issue by ruffian** [thoukydides/homebridge-homeconnect#199]
### Plugin Version

0.37.0

### Home Connect Appliance(s)

Bosch Washer WAVH8E41CH/07

### Log File

```shell
▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼▼
[12/31/2023, 10:28:49 AM] [HomeConnect] 
[12/31/2023, 10:28:49 AM] [HomeConnect] // Union types
[12/31/2023, 10:28:49 AM] [HomeConnect] export type ProgramKey =
[12/31/2023, 10:28:49 AM] [HomeConnect]     'ConsumerProducts.CoffeeMaker.Program.Beverage.CaffeLatte'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.Cappuccino'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.Coffee'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.Espresso'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.EspressoMacchiato'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.HotWater'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.LatteMacchiato'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.MilkFroth'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.Ristretto'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.Beverage.WarmMilk'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Americano'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.CafeAuLait'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.CafeConLeche'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.CafeCortado'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Cortado'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Doppio'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.FlatWhite'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Galao'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Garoto'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.GrosserBrauner'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Kaapi'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.KleinerBrauner'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.KoffieVerkeerd'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.RedEye'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Verlaengerter'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.VerlaengerterBraun'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.WienerMelange'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.BottomHeating'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.FrozenHeatupSpecial'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.HotAir'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.HotAirEco'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.HotAirGrilling'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.KeepWarm'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.PizzaSetting'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.PreheatOvenware'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.SlowCook'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.TopBottomHeating'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Cooking.Oven.Program.HeatingMode.TopBottomHeatingEco'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.Auto2'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.Eco50'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.Glas40'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.Intensiv70'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.Kurz60'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.MachineCare'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.NightWash'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.PreRinse'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'Dishcare.Dishwasher.Program.Quick45'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'LaundryCare.Washer.Program.Auto40'
[12/31/2023, 10:28:49 AM] [HomeConnect]   | 'LaundryCare.Washer.Program.DrumClean'; // (unrecognised)
[12/31/2023, 10:28:49 AM] [HomeConnect] export type SpinSpeed =
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.EnumType.SpinSpeed.Auto'; // (unrecognised)
[12/31/2023, 10:28:49 AM] [HomeConnect] 
[12/31/2023, 10:28:49 AM] [HomeConnect] // Option
[12/31/2023, 10:28:49 AM] [HomeConnect] export interface OptionValues {
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.EnergyForecast'?:                   number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.EstimatedTotalProgramTime'?:        number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.FinishInRelative'?:                 number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.ProgramProgress'?:                  number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.RemainingProgramTime'?:             number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.RemainingProgramTimeIsEstimated'?:  boolean;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.StartInRelative'?:                  number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'BSH.Common.Option.WaterForecast'?:                    number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'Dishcare.Dishwasher.Option.ExtraDry'?:                boolean;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'Dishcare.Dishwasher.Option.HygienePlus'?:             boolean;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'Dishcare.Dishwasher.Option.IntensivZone'?:            boolean;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'Dishcare.Dishwasher.Option.VarioSpeedPlus'?:          boolean;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Common.Option.LoadRecommendation'?:       number;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Common.Option.ProcessPhase'?:             unknown; // (unrecognised)
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Common.Option.ReferToProgram'?:           ProgramKey; // (unrecognised)
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.Option.IDos1.Active'?:             unknown; // (unrecognised)
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.Option.IDos2.Active'?:             unknown; // (unrecognised)
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.Option.RinseHold'?:                boolean;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.Option.SilentWash'?:               boolean;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.Option.SpinSpeed'?:                SpinSpeed; // (unrecognised)
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.Option.Stains'?:                   Stains;
[12/31/2023, 10:28:49 AM] [HomeConnect]     'LaundryCare.Washer.Option.Temperature'?:              Temperature;
[12/31/2023, 10:28:49 AM] [HomeConnect] }
[12/31/2023, 10:28:49 AM] [HomeConnect] 
[12/31/2023, 10:28:49 AM] [HomeConnect] ▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲
```

**Comment by thoukydides** [comment]
Thanks for the report. 

That `LaundryCare.Washer.Program.DrumClean` program, and a few of those options, are actually listed in the [Home Connect API documentation](https://api-docs.home-connect.com/programs-and-options/?#washer). Not sure how I missed them. Maybe there has been an update.

The `LaundryCare.Common.Option.ProcessPhase` and `LaundryCare.Common.Option.ReferToProgram` program options are undocumented.

Anyway, I will add them all in the next release (and review the API documentation to see whether I missed anything else).

**Comment by thoukydides** [comment]
Looking more carefully...
* `LaundryCare.Washer.Program.DrumClean` is actually already known to the plugin. The problem was that I had incorrectly defined (the undocumented program option) `LaundryCare.Common.Option.ReferToProgram` as expecting a number rather than a `ProgramKey`. I will fix that.
* `LaundryCare.Washer.EnumType.SpinSpeed.Auto` is not one of the [documented Spin Speed Options](https://api-docs.home-connect.com/programs-and-options/?#washer_spin-speed-option) so I will add that.
* `LaundryCare.Common.Option.ProcessPhase` is not documented. [`ConsumerProducts.CleaningRobot.Option.ProcessPhase`](https://api-docs.home-connect.com/events/?#program-progress-changes) is documented, but those values are clearly specific to the robot, so more information will be required to add this to the plugin.
* `LaundryCare.Washer.Option.IDos1.Active` and `LaundryCare.Washer.Option.IDos2.Active` are curious. There are documented options [`LaundryCare.Washer.Option.IDos1Active`](https://api-docs.home-connect.com/programs-and-options/?#washer_i-dos-1-active-option) and [`LaundryCare.Washer.Option.IDos2Active`](https://api-docs.home-connect.com/programs-and-options/?#washer_i-dos-2-active-option) without the extra period ("`.`") before `Active`. I will add these versions as alternatives.

**Comment by thoukydides** [comment]
All of those issues should be addressed in [v0.37.1](https://github.com/thoukydides/homebridge-homeconnect/releases/tag/v0.37.1).

Thanks again for the report.
